staging/src/k8s.io/apiextensions-apiserver/pkg/controller/apiapproval/apiapproval_controller.go

```diff
@@ -0,0 +1,262 @@
+	"k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	client "k8s.io/apiextensions-apiserver/pkg/client/clientset/internalclientset/typed/apiextensions/internalversion"
+	informers "k8s.io/apiextensions-apiserver/pkg/client/informers/internalversion/apiextensions/internalversion"
+	listers "k8s.io/apiextensions-apiserver/pkg/client/listers/apiextensions/internalversion"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+	"k8s.io/klog"
+)
+
+// ConditionController is maintaining the KubeAPIApproved condition.
+type KubeAPIApprovedConditionController struct {
+	crdClient client.CustomResourceDefinitionsGetter
+
+	crdLister listers.CustomResourceDefinitionLister
+	crdSynced cache.InformerSynced
+
+	// To allow injection for testing.
+	syncFn func(key string) error
+
+	queue workqueue.RateLimitingInterface
+
+	// last generation this controller updated the condition per CRD name (to avoid two
+	// different version of the apiextensions-apiservers in HA to fight for the right message)
+	lastSeenGenerationLock sync.Mutex
+	lastSeenGeneration     map[string]int64
+}
+
+// NewKubeAPIApprovedConditionController constructs a non-structural schema condition controller.
+func NewKubeAPIApprovedConditionController(
+	crdInformer informers.CustomResourceDefinitionInformer,
+	crdClient client.CustomResourceDefinitionsGetter,
+) *KubeAPIApprovedConditionController {
+	c := &KubeAPIApprovedConditionController{
+		crdClient:          crdClient,
+		crdLister:          crdInformer.Lister(),
+		crdSynced:          crdInformer.Informer().HasSynced,
+		queue:              workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "kube_api_approved_condition_controller"),
+		lastSeenGeneration: map[string]int64{},
+	}
+
+	crdInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    c.addCustomResourceDefinition,
+		UpdateFunc: c.updateCustomResourceDefinition,
+		DeleteFunc: c.deleteCustomResourceDefinition,
+	})
+
+	c.syncFn = c.sync
+
+	return c
+}
+
+// calculateCondition determines the new KubeAPIApproved condition
+func calculateCondition(crd *apiextensions.CustomResourceDefinition) *apiextensions.CustomResourceDefinitionCondition {
+	if !apihelpers.IsProtectedCommunityGroup(crd.Spec.Group) {
+		return nil
+	}
+
+	switch approvalState := apihelpers.GetAPIApprovalState(crd.Annotations); approvalState {
+	case apihelpers.APIApprovalInvalid:
+		return &apiextensions.CustomResourceDefinitionCondition{
+			Type:    apiextensions.KubeAPIApproved,
+			Status:  apiextensions.ConditionFalse,
+			Reason:  "Annotation",
+			Message: fmt.Sprintf("%v is not approved, the format is invalid: protected groups must have approval approval with either a URL or a reason starting with \"unapproved\", see https://github.com/kubernetes/enhancements/pull/1111", crd.Name),
```

Comment:
would be good to include the annotation name in the messages to let people know what they need to do

Reply:
could probably include that instead of the CRD name... the condition is on the CRD object, so the name is pretty implicit